clients/drcachesim/tools/schedule_stats.cpp

```diff
@@ -0,0 +1,295 @@
+        if (type_is_instr(memref.instr.type))
+            line << " pc=" << std::hex << memref.instr.addr << std::dec;
+        else if (memref.marker.type == TRACE_TYPE_MARKER) {
+            line << " " << memref.marker.marker_type
+                 << " val=" << memref.marker.marker_value;
+        }
+        line << "\n";
+        std::cerr << line.str();
+    }
+    if (memref.marker.type == TRACE_TYPE_MARKER &&
+        memref.marker.marker_type == TRACE_MARKER_TYPE_CORE_WAIT) {
+        ++shard->counters.waits;
+        if (!shard->prev_was_wait) {
+            shard->thread_sequence += '-';
+            shard->cur_segment_instrs = 0;
+            shard->prev_was_wait = true;
+        } else {
+            ++shard->cur_segment_instrs;
+            if (shard->cur_segment_instrs == knob_print_every_) {
+                shard->thread_sequence += '-';
+            }
+        }
+        return true;
+    } else
+        shard->prev_was_wait = false;
+    int64_t input = shard->stream->get_input_id();
+    if (input != shard->prev_input) {
+        // We convert to letters which only works well for <=26 inputs.
+        if (!shard->thread_sequence.empty()) {
+            ++shard->counters.total_switches;
+            if (shard->saw_maybe_blocking || shard->saw_exit)
+                ++shard->counters.voluntary_switches;
+            if (shard->direct_switch_target == memref.marker.tid)
+                ++shard->counters.direct_switches;
+            shard->thread_sequence += ',';
+        }
+        shard->thread_sequence += 'A' + static_cast<char>(input % 26);
+        shard->cur_segment_instrs = 0;
+        if (knob_verbose_ >= 2) {
+            std::ostringstream line;
+            line << "Core #" << std::setw(2) << shard->core << " @" << std::setw(9)
+                 << shard->stream->get_record_ordinal() << " refs, " << std::setw(9)
+                 << shard->stream->get_instruction_ordinal() << " instrs: input "
+                 << std::setw(4) << input << " @" << std::setw(9)
+                 << shard->stream->get_input_interface()->get_record_ordinal()
+                 << " refs, " << std::setw(9)
+                 << shard->stream->get_input_interface()->get_instruction_ordinal()
+                 << " instrs, time "
+                 << std::setw(16)
+                 // TODO i#5843: For time quanta, provide some way to get the
+                 // latest time and print that here instead of the the timestamp?
+                 << shard->stream->get_input_interface()->get_last_timestamp()
+                 << " == thread " << memref.instr.tid << "\n";
+            std::cerr << line.str();
+        }
+        shard->prev_input = input;
+    }
+    if (type_is_instr(memref.instr.type)) {
+        ++shard->counters.instrs;
+        // Print a single letter for any partial sequence.
+        if (shard->cur_segment_instrs == 0)
+            shard->thread_sequence += 'A' + static_cast<char>(input % 26);
```

Comment:
At a thread switch L186 adds the char to the thread sequence and sets cur_segment_insrs to zero, and here we'd add the char again?

Reply:
Ah good catch.  Somehow I thought it was only adding after N so you wouldn't get a letter for a shorter sequence...not sure if the code changed in between.  But now, yes, this is redundant.  Moved this append inside the `== knob_print_every_`.